### ISSUE TITLE
BIO_dgram_pair test breakage, BIO_write with zero length writes

### ISSUE THREAD

**Issue by hlandau** [openssl/openssl#19267]
We have a stochastic test failure for BIO_dgram_pair which just merged. This is caused by a call to BIO_write with a zero length buffer.
I designed BIO_dgram_pair to support zero-length datagrams. However, `bio_lib.c` filters these calls:

```c
int BIO_write(BIO *b, const void *data, int dlen)
{
    size_t written;
    int ret;

    if (dlen <= 0)
        return 0;

    ret = bio_write_intern(b, data, (size_t)dlen, &written);

    if (ret > 0) {
        /* written should always be <= dlen */
        ret = (int)written;
    }

    return ret;
}
```

This probably means we also have a "bug" in BIO_dgram as far as supporting zero-length UDP datagrams is concerned.

Our options are:

- Change `<= 0` to `< 0`
  Pros: BIO_dgram, BIO_dgram_pair can support zero-length datagrams properly
  Cons: Possible compatibility issues?
- Allow some kind of compatibility flag to be set inside the BIO by the method to allow zero-length writes
  Pros: BIO_dgram, BIO_dgram_pair can support zero-length datagrams properly
  Pros: No compatibility issues
- Decide not to support zero-length datagrams (at least via the BIO_read/BIO_write API?)

**Comment by mattcaswell** [comment]
But the BIO_dgram_pair PR added this documentation:

> This BIO does not support sending or receiving zero-length datagrams. Passing a
zero-length buffer to BIO_write is treated as a no-op.

**Comment by hlandau** [comment]
Hm, I forgot about that... anyway, that makes things simple. I'll just fix the test.

**Comment by hlandau** [comment]
Fix in #19268
